aten/src/ATen/native/ConvolutionMM2d.cpp

```diff
@@ -0,0 +1,575 @@
+
+  Tensor grad_weight_2d;
+  if (grad_weight.defined()) {
+    checkContiguous(c, grad_weight_arg);
+    grad_weight_2d = view_weight_2d(grad_weight);
+  }
+
+  if (grad_bias.defined()) {
+    checkContiguous(c, grad_bias_arg);
+  }
+
+  slow_conv2d_shape_check(
+      input_,
+      grad_output_,
+      grad_weight_2d,
+      grad_bias,
+      kernel_height,
+      kernel_width,
+      stride_height,
+      stride_width,
+      pad_height,
+      pad_width,
+      true);
+
+  auto input = input_.contiguous();
+  auto grad_output = grad_output_.contiguous();
+
+  const int64_t batch_size = input.size(0);
+  for (int64_t t = 0; t < batch_size; t++) {
+    Tensor grad_output_t = grad_output[t];
+    Tensor finput_t;
+    if (grad_weight_2d.defined()) {
+      finput_t = finput[t];
+    }
+
+    slow_conv2d_backward_parameters_frame(
+        grad_weight_2d, grad_bias, grad_output_t, finput_t);
+  }
+}
+
+} // namespace
+
+std::tuple<Tensor&, Tensor&, Tensor&> slow_conv2d_forward_out_cpu(
+    Tensor& output,
+    Tensor& finput,
+    Tensor& fgrad_input,
+    const Tensor& self,
+    const Tensor& weight_,
+    IntArrayRef kernel_size,
+    const Tensor& bias,
+    IntArrayRef stride,
+    IntArrayRef padding) {
+  const int64_t kernel_height = kernel_size[0];
+  const int64_t kernel_width = kernel_size[1];
+  const int64_t pad_height = padding[0];
+  const int64_t pad_width = padding[1];
+  const int64_t stride_height = stride[0];
+  const int64_t stride_width = stride[1];
+
+  const Tensor weight_2d = view_weight_2d(weight_);
+
+  slow_conv2d_shape_check(
+      self,
+      Tensor(),
+      weight_2d,
+      bias,
+      kernel_height,
+      kernel_width,
+      stride_height,
+      stride_width,
+      pad_height,
+      pad_width,
+      false);
+
+  const Tensor input = self.contiguous();
+  const int64_t ndim = input.dim();
+  const int64_t dim_planes = 1;
+  const int64_t dim_height = 2;
+  const int64_t dim_width = 3;
+
+  const int64_t n_input_plane = input.size(dim_planes);
+  const int64_t input_height = input.size(dim_height);
+  const int64_t input_width = input.size(dim_width);
+  const int64_t n_output_plane = weight_2d.size(0);
+  const int64_t output_height =
+      (input_height + 2 * pad_height - kernel_height) / stride_height + 1;
+  const int64_t output_width =
+      (input_width + 2 * pad_width - kernel_width) / stride_width + 1;
+
+  const int64_t batch_size = input.size(0);
+
+  finput.resize_({batch_size,
+                  n_input_plane * kernel_height * kernel_width,
+                  output_height * output_width});
+  output.resize_({batch_size, n_output_plane, output_height, output_width});
+
+  at::parallel_for(0, batch_size, 0, [&](int64_t start, int64_t end) {
+    NoGradGuard no_grad;
+    AutoNonVariableTypeMode non_variable_type_mode;
+    for (int64_t t = start; t < end; t++) {
+      Tensor input_t = input[t];
+      Tensor output_t = output[t];
+      Tensor finput_t = finput[t];
```

Comment:
@albanD Another case of Tensor operations being done inside of a loop

Reply:
cc @ilia-cher